webbpsf/wfirst.py

```diff
@@ -3,43 +3,330 @@
+        interpolator = WavelengthDependenceInterpolator(n_wavelengths=len(wavelengths),
+                                                        n_zernikes=22)
+        for row in rows:
+            z = np.zeros(22)
+            for idx in xrange(3, 22):  # omit piston, tip, tilt for WFI
+                z[idx] = row['Z{}'.format(idx + 1)]
+            interpolator.set_aberration_terms(row['Wave(um)'] * 1e-6, z)
+
+        return interpolator
+
+    detector_ids = set(zernike_table['Cnfg#'])
+    for detid in detector_ids:
+        detectors["SCA{:02}".format(detid)] = build_detector_from_table(detid, zernike_table)
+
+    return detectors
 
 class WFIRSTInstrument(webbpsf_core.SpaceTelescopeInstrument):
     """
     WFIRSTInstrument contains data and functionality common to WFIRST
     instruments, such as setting the pupil shape
+    """
+    telescope = "WFIRST"
+    _detectors = {}
+    """
+    Dictionary mapping detector names to FieldDependentAberration optics.
+
+    (Subclasses must populate this at `__init__`.)
+    """
+    _selected_detector = None
+    """
+    The name of the currently selected detector. Must be a key in _detectors, as validated by the
+    `detector` property setter.
 
-    WARNING: No realistic wavefront error map was available for WFIRST at release time.
-             This assumes a perfect telescope!
+    (Subclasses must populate this at `__init__`.)
+    """
+    _detector_position = (2048.0, 2048.0)
+    """
+    The pixel position, accessed through the `detector_position` property
+    and validated to be in-range on the selected detector.
     """
+
     def __init__(self, *args, **kwargs):
         super(WFIRSTInstrument, self).__init__(*args, **kwargs)
-        # the AFTA_symmetrical.fits pupil is 516x516, covering a 2.4m diameter
-        pixelscale = 2.4 / 516.0
-        self.pupil = poppy.FITSOpticalElement(
-            transmission=os.path.join(self._WebbPSF_basepath, 'AFTA_symmetrical.fits'),
-            pixelscale=pixelscale
-        )
-        self.pupilopd = None  # until we have some OPD maps and a FITS pupil of the right shape
+        self.pupil = os.path.join(self._WebbPSF_basepath, 'AFTA_WFC_C5_Pupil_Shortwave_Norm_2048px.fits')
+        self.pupilopd = os.path.join(self._WebbPSF_basepath, 'upscaled_HST_OPD.fits')
+
+        # n.b. WFIRSTInstrument subclasses must set these
+        self._detectors = {}
+        self._selected_detector = None
+
+    @property
+    def detector(self):
+        """Detector selected for simulated PSF
+
+        Used in calculation of field-dependent aberrations. Must be
+        selected from detectors in the `detector_list` attribute.
+        """
+        return self._selected_detector
 
-class WFIRSTImager(WFIRSTInstrument):
+    @detector.setter
+    def detector(self, value):
+        if value.upper() not in self.detector_list:
+            raise ValueError("Invalid aperture. Valid aperture names are: {}".format(', '.join(self.detector_list)))
```

Comment:
the error text here should probably say 'detector' instead of 'aperture' for local consistency?